// eyeglass/lib/util/SassImplementation.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var assertions_1 = require("./assertions");
function isSassValue(sass, value) {
    if (value && typeof value === "object") {
        return isSassNumber(sass, value)
            || isSassString(sass, value)
            || isSassColor(sass, value)
            || isSassBoolean(sass, value)
            || isSassList(sass, value)
            || isSassMap(sass, value)
            || isSassNull(sass, value);
    }
    else {
        return false;
    }
}
exports.isSassValue = isSassValue;
function inspect(sass, value) {
    if (isSassNumber(sass, value)) {
        return "" + value.getValue() + value.getUnit();
    }
    else if (isSassString(sass, value)) {
        return value.getValue();
    }
    else if (isSassColor(sass, value)) {
        return "rgba(" + value.getR() + ", " + value.getG() + ", " + value.getB() + ", " + value.getA() + ")";
    }
    else if (isSassBoolean(sass, value)) {
        if (value === sass.TRUE) {
            return "true";
        }
        else {
            return "false";
        }
    }
    else if (isSassList(sass, value)) {
        var s = "(";
        for (var i = 0; i < value.getLength(); i++) {
            if (i > 0) {
                if (value.getSeparator()) {
                    s += ", ";
                }
                else {
                    s += " ";
                }
            }
            s += inspect(sass, value.getValue(i));
        }
        s += ")";
        return s;
    }
    else if (isSassMap(sass, value)) {
        var s = "(";
        for (var i = 0; i < value.getLength(); i++) {
            if (i > 0) {
                s += ", ";
            }
            s += value.getKey(i);
            s += ": ";
            s += inspect(sass, value.getValue(i));
        }
        s += ")";
        return s;
    }
    else if (isSassNull(sass, value)) {
        return "null";
    }
    else {
        return assertions_1.unreachable();
    }
}
exports.inspect = inspect;
function isSassNumber(sass, value) {
    return value instanceof sass.types.Number;
}
exports.isSassNumber = isSassNumber;
function isSassString(sass, value) {
    return value instanceof sass.types.String;
}
exports.isSassString = isSassString;
function isSassColor(sass, value) {
    return value instanceof sass.types.Color;
}
exports.isSassColor = isSassColor;
function isSassBoolean(sass, value) {
    return value instanceof sass.types.Boolean;
}
exports.isSassBoolean = isSassBoolean;
function isSassNull(sass, value) {
    return value === sass.NULL;
}
exports.isSassNull = isSassNull;
function isSassList(sass, value) {
    return value instanceof sass.types.List;
}
exports.isSassList = isSassList;
function isSassMap(sass, value) {
    return value instanceof sass.types.Map;
}
exports.isSassMap = isSassMap;
function isSassMapOrEmptyList(sass, value) {
    return isSassMap(sass, value) || (isSassList(sass, value) && value.getLength() === 0);
}
exports.isSassMapOrEmptyList = isSassMapOrEmptyList;
function isSassError(sass, value) {
    return value instanceof sass.types.Error;
}
exports.isSassError = isSassError;
function isSassImplementation(impl) {
    return (impl && typeof impl === "object"
        && impl !== null
        && typeof impl.render === "function"
        && typeof impl.renderSync === "function"
        && typeof impl.types === "object"
        && typeof impl.info === "string");
}
exports.isSassImplementation = isSassImplementation;
var typeGuards = {
    null: isSassNull,
    string: isSassString,
    number: isSassNumber,
    map: isSassMapOrEmptyList,
    list: isSassList,
    color: isSassColor,
    boolean: isSassBoolean,
    error: isSassError,
};
function typeName(sass, value) {
    if (isSassNull(sass, value))
        return "null";
    if (isSassString(sass, value))
        return "string";
    if (isSassNumber(sass, value))
        return "number";
    if (isSassMap(sass, value))
        return "map";
    if (isSassList(sass, value))
        return "list";
    if (isSassColor(sass, value))
        return "color";
    if (isSassBoolean(sass, value))
        return "boolean";
    if (isSassError(sass, value))
        return "error";
    return assertions_1.unreachable(value);
}
function isType(sass, value, name) {
    var guard = typeGuards[name];
    if (guard(sass, value)) {
        return true;
    }
    else {
        return false;
    }
}
exports.isType = isType;
function typeError(sass, expected, actual) {
    return sass.types.Error("Expected " + expected + ", got " + (typeof actual === "string" ? actual : typeName(sass, actual)) + (typeof actual === "string" ? "" : ": " + inspect(sass, actual)));
}
exports.typeError = typeError;
var SassTypeError = /** @class */ (function (_super) {
    __extends(SassTypeError, _super);
    function SassTypeError(sass, expected, actual) {
        return _super.call(this, "Expected " + expected + ", got " + (typeof actual === "string" ? actual : typeName(sass, actual)) + (typeof actual === "string" ? "" : ": " + inspect(sass, actual))) || this;
    }
    return SassTypeError;
}(Error));
exports.SassTypeError = SassTypeError;
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
function helpers(sass) {
    return {
        isSass: function (value) { return isSassImplementation(value); },
        isNull: function (value) { return isSassNull(sass, value); },
        isString: function (value) { return isSassString(sass, value); },
        isNumber: function (value) { return isSassNumber(sass, value); },
        isMap: function (value) { return isSassMap(sass, value); },
        isList: function (value) { return isSassList(sass, value); },
        isColor: function (value) { return isSassColor(sass, value); },
        isBoolean: function (value) { return isSassBoolean(sass, value); },
        isError: function (value) { return isSassError(sass, value); },
        isValue: function (value) { return isSassValue(sass, value); },
        isMapOrEmptyList: function (value) { return isSassMapOrEmptyList(sass, value); },
        typeError: typeError.bind(null, sass),
        isType: isType.bind(null, sass),
        typeName: typeName.bind(null, sass),
        inspect: inspect.bind(null, sass),
        TypeError: /** @class */ (function (_super) {
            __extends(class_1, _super);
            function class_1(expected, actual) {
                return _super.call(this, sass, expected, actual) || this;
            }
            return class_1;
        }(SassTypeError))
    };
}
exports.helpers = helpers;
